### ISSUE TITLE
"test"  compare with "predict"  have different result for the same image

### ISSUE THREAD

**Issue by feiyang2010jin** [emedvedev/attention-ocr#86]
@emedvedev 
hello   thank you for your share !   I Meet a problem when test the model,
why for the same image , I chose "test"  and "predict"  recognition , but return different result.

**Comment by emedvedev** [comment]
Hi! It would be great to have more information on this: full logs, the reference image (and ideally the checkpoint), information about your dataset. That'll give me something to work with.

Another question: are you getting _consistent_ results with both `test` and `predict`? Do they always differ in the same way?

**Comment by feiyang2010jin** [comment]
2018-05-11 14:32:29.978380: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:02:00.0, compute capability: 6.1)
2018-05-11 14:32:30,183 root  INFO     phase: predict
2018-05-11 14:32:30,183 root  INFO     model_dir: ./checkpoints
2018-05-11 14:32:30,183 root  INFO     load_model: True
2018-05-11 14:32:30,184 root  INFO     output_dir: ./results
2018-05-11 14:32:30,184 root  INFO     steps_per_checkpoint: 100
2018-05-11 14:32:30,184 root  INFO     batch_size: 65
2018-05-11 14:32:30,184 root  INFO     learning_rate: 65.000000
2018-05-11 14:32:30,184 root  INFO     reg_val: 0
2018-05-11 14:32:30,184 root  INFO     max_gradient_norm: 5.000000
2018-05-11 14:32:30,184 root  INFO     clip_gradients: True
2018-05-11 14:32:30,184 root  INFO     max_image_width 500.000000
2018-05-11 14:32:30,185 root  INFO     max_prediction_length 20.000000
2018-05-11 14:32:30,185 root  INFO     channels: 1
2018-05-11 14:32:30,185 root  INFO     target_embedding_size: 10.000000
2018-05-11 14:32:30,185 root  INFO     attn_num_hidden: 128
2018-05-11 14:32:30,185 root  INFO     attn_num_layers: 2
2018-05-11 14:32:30,185 root  INFO     visualize: True
2018-05-11 14:32:34,125 root  INFO     Reading model parameters from ./checkpoints/model.ckpt-335200
result: ok 0.18 LLGLLAA2650010275
/data/wangchongjin/VINTest/test1/LJDDAA22650002375{split}2795.jpg LLGLLAA2650010275



but when execute " test "  it results  correct!


2018-05-11 15:16:14,342 root  INFO     phase: test
2018-05-11 15:16:14,342 root  INFO     model_dir: ./checkpoints
2018-05-11 15:16:14,342 root  INFO     load_model: True
2018-05-11 15:16:14,342 root  INFO     output_dir: ./results
2018-05-11 15:16:14,342 root  INFO     steps_per_checkpoint: 100
2018-05-11 15:16:14,342 root  INFO     batch_size: 1
2018-05-11 15:16:14,342 root  INFO     learning_rate: 65.000000
2018-05-11 15:16:14,343 root  INFO     reg_val: 0
2018-05-11 15:16:14,343 root  INFO     max_gradient_norm: 5.000000
2018-05-11 15:16:14,343 root  INFO     clip_gradients: True
2018-05-11 15:16:14,343 root  INFO     max_image_width 500.000000
2018-05-11 15:16:14,343 root  INFO     max_prediction_length 20.000000
2018-05-11 15:16:14,343 root  INFO     channels: 1
2018-05-11 15:16:14,343 root  INFO     target_embedding_size: 10.000000
2018-05-11 15:16:14,343 root  INFO     attn_num_hidden: 128
2018-05-11 15:16:14,343 root  INFO     attn_num_layers: 2
2018-05-11 15:16:14,343 root  INFO     visualize: True
2018-05-11 15:16:18,097 root  INFO     Reading model parameters from ./checkpoints/model.ckpt-335200
2018-05-11 15:16:20.422495: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:02:00.0, compute capability: 6.1)
2018-05-11 15:16:29,253 root  INFO     Step 1 (6.557s). Accuracy: 100.00%, loss: 0.001158, perplexity: 1.00116, probability: 97.92% 100% (LJDDAA22650002375)
correct 1.000000  1  1

**Comment by martinbacsi** [comment]
I'm experiencing the same issue

**Comment by emedvedev** [comment]
Which version are you trying? Is it the latest `pip` or the `master` branch? If it's the `master`, can you please try again?

**Comment by martinbacsi** [comment]
I was using the pip version (0.7.4) now installed the master branch. still experiencing the issue.

**Comment by Harmonicahappy** [comment]
> Which version are you trying? Is it the latest `pip` or the `master` branch? If it's the `master`, can you please try again?

I think you might have written a bug. In your test mode, you use `def step(self, batch, forward_only)` function, so decoder_inputs will be filled with ground truth labels, but the decoder should use its softmax output at time T-1 as its decoder_inputs at time T in inference mode.

**Comment by emedvedev** [comment]
Oh, thanks for the report. Could you try implementing it on your local copy and submit a PR if you manage to improve the result / fix the problem? I’d really appreciate it!

**Comment by Harmonicahappy** [comment]
> Oh, thanks for the report. Could you try implementing it on your local copy and submit a PR if you manage to improve the result / fix the problem? I’d really appreciate it!
> […](#)
> On Apr 11, 2019, 20:49 -0700, Terence Wu ***@***.***>, wrote: > Which version are you trying? Is it the latest pip or the master branch? If it's the master, can you please try again? I think you might have written a bug. In your test mode, you use def step(self, batch, forward_only) function, so decoder_inputs will be filled with ground truth labels, but the decoder should use its softmax output at time T-1 as its decoder_inputs at time T in inference mode. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

Sure, I will try to fix it in about 2 weeks because I am still not clear about some details.

**Comment by emedvedev** [comment]
Fixed in https://github.com/emedvedev/attention-ocr/pull/132
